src/AdvancedCircuitBreaker.php

```diff
@@ -0,0 +1,145 @@
+<?php
+/**
+ * 2007-2019 PrestaShop SA and Contributors
+ *
+ * NOTICE OF LICENSE
+ *
+ * This source file is subject to the Open Software License (OSL 3.0)
+ * that is bundled with this package in the file LICENSE.txt.
+ * It is also available through the world-wide-web at this URL:
+ * https://opensource.org/licenses/OSL-3.0
+ * If you did not receive a copy of the license and are unable to
+ * obtain it through the world-wide-web, please send an email
+ * to license@prestashop.com so we can send you a copy immediately.
+ *
+ * DISCLAIMER
+ *
+ * Do not edit or add to this file if you wish to upgrade PrestaShop to newer
+ * versions in the future. If you wish to customize PrestaShop for your
+ * needs please refer to http://www.prestashop.com for more information.
+ *
+ * @author    PrestaShop SA <contact@prestashop.com>
+ * @copyright 2007-2019 PrestaShop SA and Contributors
+ * @license   https://opensource.org/licenses/OSL-3.0 Open Software License (OSL 3.0)
+ * International Registered Trademark & Property of PrestaShop SA
+ */
+
+namespace PrestaShop\CircuitBreaker;
+
+use PrestaShop\CircuitBreaker\Contracts\Client;
+use PrestaShop\CircuitBreaker\Contracts\Storage;
+use PrestaShop\CircuitBreaker\Contracts\System;
+use PrestaShop\CircuitBreaker\Contracts\Transitioner;
+use PrestaShop\CircuitBreaker\Exceptions\UnavailableServiceException;
+
+/**
+ * This implementation of the CircuitBreaker is a bit more advanced than the SimpleCircuitBreaker,
+ * it allows you to setup your client, system or storage. You can also add an optional transitioner.
+ * And you can send requests with parameters.
+ */
+class AdvancedCircuitBreaker extends PartialCircuitBreaker implements Transitioner
+{
+    /** @var Transitioner */
+    protected $transitioner;
+
+    /**
+     * @param System $system
+     * @param Client $client
+     * @param Storage $storage
+     * @param Transitioner|null $transitioner
+     */
+    public function __construct(System $system, Client $client, Storage $storage, Transitioner $transitioner = null)
```

Comment:
I don't think `$transitioner` should be nullable. If need be, you could pass along a `NullTransitioner`.

Reply:
Done